### ISSUE TITLE
Your client is using unsupported sync protocol (10, supported version: 9) with latest version of Anki

### ISSUE THREAD

**Issue by kalehmann** [ankicommunity/anki-sync-server#13]
## How to reproduce ?

- First I set up the latest master of the anki-sync-server
- Then I go to [apps.ankiweb.net](https://apps.ankiweb.net/) and download the latest stable release (2.1.22).
- After that I follow the instructions in the [Readme](https://github.com/ankicommunity/anki-sync-server#setting-up-anki) to setup my sync server in Anki.
- Now I hit synchronize

## Expected behavior

Everything synchronizes just fine

## Actual behavior
I get a dialog with the message

> Your client is using unsupported sync protocol (10, supported version: 9)

## Additional information

The message is issued in [sync_app.py#L90](https://github.com/ankicommunity/anki-sync-server/blob/c6205a1c438ad338450f09462d60005964cc552e/ankisyncd/sync_app.py#L90)

**Comment by VikashKothary** [comment]
Hi, thanks for raising this issue. I've had a look at this issue.

## TL;DR;

Currently Anki 2.1.22 is **not officially supported** by Anki Sync Server as stated in the README. That being said, there have been users who've patched it to make it work.
This needs considerable testing before it will officially be supported.

1. Commenting lines 89 and 90 will allow the basic sync to work. 
2. Updating `__init__.py` should fix media sync.
```
import anki.sync, anki.hooks, aqt

addr = "http://127.0.0.1:27701/"
anki.sync.SYNC_BASE = "%s" + addr
aqt.mediasync.SYNC_BASE = "%s" + addr

def resetHostNum():
    aqt.mw.pm.profile['hostNum'] = None

anki.hooks.addHook("profileLoaded", resetHostNum)
```

NOTE: You will still get an error `AnkiWeb ID or password was incorrect; please try again`. You can ignore this. The sync should have worked. See [this notebook](https://github.com/ankicommunity/anki-sync-server/blob/develop/notebooks/read_collections.ipynb) to see how to read your collections.

**Comment by VikashKothary** [comment]
## More information for Official Fix

The problem as you stated is that the [sync_app.py#L90](https://github.com/ankicommunity/anki-sync-server/blob/c6205a1c438ad338450f09462d60005964cc552e/ankisyncd/sync_app.py#L90) is blocking the sync. It is doing this because the sync version (SYN_VER) in the Anki Sync Server is 9 while the one used by your client is 10.

To fix this, we'd need to update the anki submodule to any commit after https://github.com/ankitects/anki/commit/131d37dca52c29033432e0149052093ab1c79461. But this is easier said than done because there has been multiple project structures changes. In the latest version of the module, the `SYNC_VER` constant is no longer available.

A change in the sync version is likely to indicate a breaking change. I don't quite know what this change it. Testing is needed to check which features may have been broken. Basic trial and error using the above fix does not show any issues as of yet.

**Comment by kalehmann** [comment]
Hi, thanks for the input. 

>  To fix this, we'd need to update the anki submodule to any commit after ankitects/anki@131d37d. But this is easier said than done because there has been multiple project structures changes. In the latest version of the module, the SYNC_VER constant is no longer available.

I am still interested in patching the code to work with the latest version of the anki codebase. I will update this issue in the next few days with a summary of all the components needed by the sync server, that have moved form the python to the rust codebase.

Hopefully I can then provide a pull request that works with the latest version of anki. In my opinion this is necessary for this project to survive in the long term.

**Comment by VikashKothary** [comment]
Hi @kalehmann,

That would be amazing if you could. I totally agree with you in the keeping this project in sync with the latest versions of Anki client is key to the success of this project.

I've created a new issue for the upgrade (see https://github.com/ankicommunity/anki-sync-server/issues/16). I'd be interested to see what you come up with.

**Comment by bartowski-l** [comment]
i have a problem desktop app 2.1.22 always show user or passwd wrong and retry . the others is be ok ,but this session when i shut the app it will show . i can't find how to fix it?who can help me ? thank you

**Comment by VikashKothary** [comment]
Hi @bartowski-l, 

So this is an known issue that hasn't been fixed as it does not stop the actual sync functionality of the sync server. See below:
> NOTE: You will still get an error `AnkiWeb ID or password was incorrect; please try again`. You can ignore this. The sync should have worked. See [this notebook](https://github.com/ankicommunity/anki-sync-server/blob/develop/notebooks/read_collections.ipynb) to see how to read your collections.

I know it may be annoying but officially the latest clients aren't supported. A possible fix, if you don't want to see the message, maybe to use an older Anki client. See below:
> The currently supported version range is 2.1.1〜2.1.11, with the exception of 2.1.9

Hopefully this helps.

**Comment by kalehmann** [comment]
Hi @bartowski-l,
I am still working on patching anki-sync-server to work with the latest version of Anki.  During this work I will also try to patch the message you are mentioning.

**Comment by kalehmann** [comment]
I am making progress on this issue. My current target is to get the sync-server working with Anki 2.1.26, **without any** client side errors or notifications.

Newer versions of Anki are currently excluded because of #19.

**Comment by AntonOfTheWoods** [comment]
> Newer versions of Anki are currently excluded because of #19.

It turns out #19 is a bit of a non-issue (see my comment to that issue) - the addon will do less and it will be a little less flexible but it will be quite easy to create platform-dependent application launch links so that you can point to a custom sync server.

**Comment by kalehmann** [comment]
I have finished my work and provided a pull request for the changes: #20.

**Comment by kalehmann** [comment]
As #20 is now merged, this issue can be closed

**Comment by update-freak** [comment]
Is the docker container really up to date on the URL [kuklinistvan/anki-sync-server:latest](https://hub.docker.com/r/kuklinistvan/anki-sync-server) ?
I got the same error measage.

**Comment by kalehmann** [comment]
I moved away from this anki sync server implementation, so I probably won't be able to help you out much.

However, I checked the image you provided and it seems like the contents have been last updated in February or August of 2020:

```
$ docker pull kuklinistvan/anki-sync-server
$ docker run -it --entrypoint=/bin/sh kuklinistvan/anki-sync-server
/app/anki-sync-server # ls -lah
total 68
drwxr-xr-x    1 root     root         254 Aug 26  2020 .
drwxr-xr-x    1 root     root          66 Aug 26  2020 ..
-rw-r--r--    1 root     root          51 Feb  6  2020 .git
-rw-r--r--    1 root     root          54 Feb  6  2020 .gitignore
-rw-r--r--    1 root     root          87 Feb  6  2020 .gitmodules
-rw-r--r--    1 root     root       33.7K Feb  6  2020 COPYING
-rw-r--r--    1 root     root        6.4K Feb  6  2020 README.md
drwxr-xr-x    1 root     root          44 Feb  6  2020 addon
drwxr-xr-x    1 root     root         440 Feb  6  2020 anki-bundled
-rwxr-xr-x    1 root     root        2.1K Feb  6  2020 ankisyncctl.py
drwxr-xr-x    1 root     root         190 Feb  6  2020 ankisyncd
lrwxrwxrwx    1 root     root          24 Aug 26  2020 ankisyncd.conf -> /app/data/ankisyncd.conf
-rw-r--r--    1 root     root         884 Feb  6  2020 ankisyncd.conf.example
drwxr-xr-x    1 root     root         386 Feb  6  2020 tests
drwxr-xr-x    1 root     root          44 Feb  6  2020 utils
```

Since my pull request #20 has been merged in September 2020, my changes are missing in the image you provided.

However, you can build your own docker image from the current state of the repository at any time with:

```
$ git clone https://github.com/ankicommunity/anki-sync-server.git
$ cd anki-sync-server
$ docker build -t anki-sync-server .
```

**Comment by update-freak** [comment]
Thank you with your idea to create it via the commands.
I tried it and got this error message when I try to login via AnkiDroid:

[log.txt](https://github.com/ankicommunity/anki-sync-server/files/13904025/log.txt)

**Comment by LuckyTurtleDev** [comment]
see also https://github.com/ankicommunity/anki-sync-server/issues/158
